### PR TITLE
3scale 2.14 release candidate

### PR DESCRIPTION
**Pull request by eguzki**
### Compare bundles: this PR's (v0.11.0) and [0.10.1](https://github.com/redhat-openshift-ecosystem/community-operators-prod/tree/main/operators/3scale-community-operator/0.10.1)

```
diff -r 0.10.1/ 0.11.0/
```

Result: https://pastebin.com/07PRhWaH

Changes that cannot be added in this repo and should be added in 
https://github.com/redhat-openshift-ecosystem/community-operators-prod

```diff
diff --git a/bundle/manifests/3scale-operator.clusterserviceversion.yaml b/bundle/manifests/3scale-operator.clusterserviceversion.yaml
index 0155950d..d451478d 100644
--- a/bundle/manifests/3scale-operator.clusterserviceversion.yaml
+++ b/bundle/manifests/3scale-operator.clusterserviceversion.yaml
@@ -217,7 +217,7 @@ metadata:
     operatorframework.io/arch.amd64: supported
     operatorframework.io/arch.ppc64le: supported
     operatorframework.io/arch.s390x: supported
-  name: 3scale-operator.v0.11.0
+  name: 3scale-community-operator.v0.11.0
   namespace: placeholder
 spec:
   apiservicedefinitions: {}
diff --git a/bundle/metadata/annotations.yaml b/bundle/metadata/annotations.yaml
index a4ab00c0..7ace76b5 100644
--- a/bundle/metadata/annotations.yaml
+++ b/bundle/metadata/annotations.yaml
@@ -4,9 +4,10 @@ annotations:
   operators.operatorframework.io.bundle.manifests.v1: manifests/
   operators.operatorframework.io.bundle.mediatype.v1: registry+v1
   operators.operatorframework.io.bundle.metadata.v1: metadata/
-  operators.operatorframework.io.bundle.package.v1: 3scale-operator
+  operators.operatorframework.io.bundle.package.v1: 3scale-community-operator
   operators.operatorframework.io.metrics.builder: operator-sdk-v1.2.0
   operators.operatorframework.io.metrics.mediatype.v1: metrics+v1
   operators.operatorframework.io.metrics.project_layout: go.kubebuilder.io/v2
   operators.operatorframework.io.test.config.v1: tests/scorecard/
   operators.operatorframework.io.test.mediatype.v1: scorecard+v1
+  com.redhat.openshift.versions: v4.11-v4.15
```

### Verification step: Fresh install

* Apply patch to complete the bundle

```diff
cat << EOF | git apply
diff --git a/bundle/manifests/3scale-operator.clusterserviceversion.yaml b/bundle/manifests/3scale-operator.clusterserviceversion.yaml
index 0155950d..d451478d 100644
--- a/bundle/manifests/3scale-operator.clusterserviceversion.yaml
+++ b/bundle/manifests/3scale-operator.clusterserviceversion.yaml
@@ -217,7 +217,7 @@ metadata:
     operatorframework.io/arch.amd64: supported
     operatorframework.io/arch.ppc64le: supported
     operatorframework.io/arch.s390x: supported
-  name: 3scale-operator.v0.11.0
+  name: 3scale-community-operator.v0.11.0
   namespace: placeholder
 spec:
   apiservicedefinitions: {}
diff --git a/bundle/metadata/annotations.yaml b/bundle/metadata/annotations.yaml
index a4ab00c0..7ace76b5 100644
--- a/bundle/metadata/annotations.yaml
+++ b/bundle/metadata/annotations.yaml
@@ -4,9 +4,10 @@ annotations:
   operators.operatorframework.io.bundle.manifests.v1: manifests/
   operators.operatorframework.io.bundle.mediatype.v1: registry+v1
   operators.operatorframework.io.bundle.metadata.v1: metadata/
-  operators.operatorframework.io.bundle.package.v1: 3scale-operator
+  operators.operatorframework.io.bundle.package.v1: 3scale-community-operator
   operators.operatorframework.io.metrics.builder: operator-sdk-v1.2.0
   operators.operatorframework.io.metrics.mediatype.v1: metrics+v1
   operators.operatorframework.io.metrics.project_layout: go.kubebuilder.io/v2
   operators.operatorframework.io.test.config.v1: tests/scorecard/
   operators.operatorframework.io.test.mediatype.v1: scorecard+v1
+  com.redhat.openshift.versions: v4.11-v4.15
EOF
```
* Build bundle image

```
make bundle-build BUNDLE_IMG=quay.io/3scale/3scale-operator:bundle-0.11.0-candidate1
```
* Push to registry
```
make bundle-image-push BUNDLE_IMG=quay.io/3scale/3scale-operator:bundle-0.11.0-candidate1
```

* Deploy the 3scale Operator in the bundle format with OLM
```sh
make bundle-run BUNDLE_IMG=quay.io/3scale/3scale-operator:bundle-0.11.0-candidate1
```

* Delete the operator (from the UI OperatorHub -> Installed Operators) and install it namespace scoped

* Deploy 3scale
```sh
CLUSTER_DOMAIN=$(oc get routes console -n openshift-console -o json | jq -r '.status.ingress[0].routerCanonicalHostname' | sed 's/router-default.//')
CURRENT_NAMESPACE=$(kubectl config current-context | cut -d / -f 1)
```
```yaml
oc apply -f - <<EOF
---
apiVersion: v1
kind: Secret
metadata:
  creationTimestamp: null
  name: aws-auth
stringData:
  AWS_ACCESS_KEY_ID: something
  AWS_SECRET_ACCESS_KEY: something
  AWS_BUCKET: something
  AWS_REGION: us-east-1
type: Opaque
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: apimanager1
spec:
  system:
    fileStorage: 
      simpleStorageService:
        configurationSecretRef:
          name: aws-auth
  wildcardDomain: "$CURRENT_NAMESPACE.$CLUSTER_DOMAIN"
  resourceRequirementsEnabled: false
EOF
```
* The smoke test passes when the APIManager status reports as *Available*
```sh
oc wait --for=condition=available apimanager/apimanager1 --timeout=-1s
```

### Verification step: Upgrade from 0.10.1 (2.13)

* Apply patch to complete the bundle

```diff
cat << EOF | git apply
diff --git a/bundle/manifests/3scale-operator.clusterserviceversion.yaml b/bundle/manifests/3scale-operator.clusterserviceversion.yaml
index 0155950d..d451478d 100644
--- a/bundle/manifests/3scale-operator.clusterserviceversion.yaml
+++ b/bundle/manifests/3scale-operator.clusterserviceversion.yaml
@@ -217,7 +217,7 @@ metadata:
     operatorframework.io/arch.amd64: supported
     operatorframework.io/arch.ppc64le: supported
     operatorframework.io/arch.s390x: supported
-  name: 3scale-operator.v0.11.0
+  name: 3scale-community-operator.v0.11.0
   namespace: placeholder
 spec:
   apiservicedefinitions: {}
diff --git a/bundle/metadata/annotations.yaml b/bundle/metadata/annotations.yaml
index a4ab00c0..7ace76b5 100644
--- a/bundle/metadata/annotations.yaml
+++ b/bundle/metadata/annotations.yaml
@@ -4,9 +4,10 @@ annotations:
   operators.operatorframework.io.bundle.manifests.v1: manifests/
   operators.operatorframework.io.bundle.mediatype.v1: registry+v1
   operators.operatorframework.io.bundle.metadata.v1: metadata/
-  operators.operatorframework.io.bundle.package.v1: 3scale-operator
+  operators.operatorframework.io.bundle.package.v1: 3scale-community-operator
   operators.operatorframework.io.metrics.builder: operator-sdk-v1.2.0
   operators.operatorframework.io.metrics.mediatype.v1: metrics+v1
   operators.operatorframework.io.metrics.project_layout: go.kubebuilder.io/v2
   operators.operatorframework.io.test.config.v1: tests/scorecard/
   operators.operatorframework.io.test.mediatype.v1: scorecard+v1
+  com.redhat.openshift.versions: v4.11-v4.15
EOF
```
* Build bundle image

```
make bundle-build BUNDLE_IMG=quay.io/3scale/3scale-operator:bundle-0.11.0-candidate1
```
* Push to registry
```
make bundle-image-push BUNDLE_IMG=quay.io/3scale/3scale-operator:bundle-0.11.0-candidate1
```

* Build custom catalog
> `opm index add` is deprecated (sqlite based catalogs are deprecated)

```
mkdir operatorindex 
cd operatorindex/
```
```
opm generate dockerfile index-dir
```

* Render existing catalog v4.15
```
opm render registry.redhat.io/redhat/community-operator-index:v4.15 -o yaml > index-dir/catalog.yaml
```
> I am testing on OCP 4.15

* add new bundle to the existing catalog
```
opm render quay.io/3scale/3scale-operator:bundle-0.11.0-candidate1 -o yaml >> index-dir/catalog.yaml
```
* Add channel entry 
```yaml
cat << EOF >> index-dir/catalog.yaml
---
schema: olm.channel
package: 3scale-community-operator
name: threescale-2.14
entries:
  - name: 3scale-community-operator.v0.11.0
    replaces: 3scale-community-operator.v0.10.1
EOF
```

* validate the catalog
```
opm validate index-dir
```

* Build the catalog
```
docker build . -f index-dir.Dockerfile -t quay.io/3scale/3scale-operator:index-with-0.11.0
```

* Push the catalog
```
docker push quay.io/3scale/3scale-operator:index-with-0.11.0
```
* Create catalogsource
```yaml
oc apply -f - << EOF
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: eguzki-catalog
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/3scale/3scale-operator:index-with-0.11.0
  displayName: Eguzki Operator Catalog
  publisher: Eguzki
EOF
```
* Install the operator on channel 2.13 from the UI OperatorHub

* Deploy 3scale 0.10.1 (2.13)
```sh
CLUSTER_DOMAIN=$(oc get routes console -n openshift-console -o json | jq -r '.status.ingress[0].routerCanonicalHostname' | sed 's/router-default.//')
CURRENT_NAMESPACE=$(kubectl config current-context | cut -d / -f 1)
```

```yaml
oc apply -f - <<EOF
---
apiVersion: v1
kind: Secret
metadata:
  creationTimestamp: null
  name: aws-auth
stringData:
  AWS_ACCESS_KEY_ID: something
  AWS_SECRET_ACCESS_KEY: something
  AWS_BUCKET: something
  AWS_REGION: us-east-1
type: Opaque
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata:
  name: apimanager1
spec:
  system:
    fileStorage: 
      simpleStorageService:
        configurationSecretRef:
          name: aws-auth
  wildcardDomain: "$CURRENT_NAMESPACE.$CLUSTER_DOMAIN"
  resourceRequirementsEnabled: false
EOF
```
* Wait until  APIManager status reports as *Available*
```
oc wait --for=condition=available apimanager/apimanager1 --timeout=-1s
```

* Check 3scale components image to verify 3scale 2.13 is installed
```
$ oc get dc
Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+
NAME                 REVISION   DESIRED   CURRENT   TRIGGERED BY
apicast-production   1          1         1         config,image(amp-apicast:2.13)
apicast-staging      1          1         1         config,image(amp-apicast:2.13)
backend-cron         1          1         1         config,image(amp-backend:2.13)
backend-listener     1          1         1         config,image(amp-backend:2.13)
backend-redis        1          1         1         config,image(backend-redis:2.13)
backend-worker       1          1         1         config,image(amp-backend:2.13)
system-app           1          1         1         config,image(amp-system:2.13)
system-memcache      1          1         1         config,image(system-memcached:2.13)
system-mysql         1          1         1         config,image(system-mysql:2.13)
system-redis         1          1         1         config,image(system-redis:2.13)
system-sidekiq       1          1         1         config,image(amp-system:2.13)
system-sphinx        1          1         1         config,image(amp-system:2.13)
zync                 1          1         1         config,image(amp-zync:2.13)
zync-database        1          1         1         config,image(zync-database-postgresql:2.13)
zync-que             1          1         1         config,image(amp-zync:2.13)
```

* Run Upgrade. From the UI OperatorHub -> Installed Operator -> 3scale API Management operator -> Subscription -> update channel to threescale-2.14

* Wait until  APIManager status reports as *Available*
```
oc wait --for=condition=available apimanager/apimanager1 --timeout=-1s
```

* Check 3scale components image to verify that 3scale has been upgraded to 2.14
```
oc get dc
Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+
NAME                 REVISION   DESIRED   CURRENT   TRIGGERED BY
apicast-production   3          1         1         config,image(amp-apicast:2.14)
apicast-staging      3          1         1         config,image(amp-apicast:2.14)
backend-cron         4          1         1         config,image(amp-backend:2.14)
backend-listener     4          1         1         config,image(amp-backend:2.14)
backend-redis        3          1         1         config,image(backend-redis:2.14)
backend-worker       3          1         1         config,image(amp-backend:2.14)
system-app           2          1         1         config,image(amp-system:2.14)
system-memcache      3          1         1         config,image(system-memcached:2.14)
system-mysql         3          1         1         config,image(system-mysql:2.14)
system-redis         3          1         1         config,image(system-redis:2.14)
system-searchd       1          1         1         config,image(system-searchd:2.14)
system-sidekiq       3          1         1         config,image(amp-system:2.14)
zync                 4          1         1         config,image(amp-zync:2.14)
zync-database        3          1         1         config,image(zync-database-postgresql:2.14)
zync-que             2          1         1         config,image(amp-zync:2.14)
```

### CODE REVIEW

File: bundle/metadata/annotations.yaml
Comment (MStokluska) [inline bundle/metadata/annotations.yaml:6]: @eguzki want to tweak this value: 
```
operators.operatorframework.io.bundle.package.v1: 3scale-operator
```
here or at the community operators repo level?

**Comment by MStokluska** [comment]
just 2 comments - apart from that it looks good to me.

File: bundle/metadata/annotations.yaml
Comment (MStokluska) [reply bundle/metadata/annotations.yaml:6]: line 7 of this file btw.

File: config/manager/manager.yaml
Comment (eguzki) [reply config/manager/manager.yaml:24]: Those were added for the downstream, so another RH product could track all the RH operators. No harm having them though. I was being consistent with previous community version that did not have those labels.

File: bundle/metadata/annotations.yaml
Comment (eguzki) [reply bundle/metadata/annotations.yaml:6]: That is part of the patch that will be applied in https://github.com/redhat-openshift-ecosystem/community-operators-prod

I do not know how to apply to upstream. We would need a hard refactor that would imply changes in downstream as well. So good enough for now.

Comment (MStokluska) [reply bundle/metadata/annotations.yaml:6]: I think we can do this upstream once we move to newer operator-sdk (it supports providing package name when building a bundle)

Comment (valerymo) [reply bundle/metadata/annotations.yaml:6]: In community PR we have correct values, similar to v0.10.1 - 
https://github.com/redhat-openshift-ecosystem/community-operators-prod/pull/4150/files#diff-bb64d4f16f07ef63aea32a8fa54c9cde707db9252f3961ee06f174cdc397b278

**Comment by valerymo** [comment]
Hi @eguzki, There are few question, I'd like to ask:
1. Are these changes a part of regular Community Release process?
2. What  is the goal of current PR ? Is it to allow All tests to pass in temporary Testing PR https://github.com/3scale/3scale-operator/pull/950 ? 
3. Is it ok if Product Release branch `origin/3scale-2.14-stable` will have changes from Community release ? 
- Could merge of current PR  impact on future Patch releases of 3scale-2.14 ?
4. Should we restore `origin/3scale-2.14-stable` branch - remove changes of Community?  (after PR 950 pass) 
5. Just for notes - All related changes are already long time in Community release PR https://github.com/redhat-openshift-ecosystem/community-operators-prod/pull/4150 . So suppose that answer of quesion #2 - Yes (?)

File: bundle/manifests/3scale-operator.clusterserviceversion.yaml
Comment (valerymo) [reply bundle/manifests/3scale-operator.clusterserviceversion.yaml:220]: thank you. ok.

File: bundle/metadata/annotations.yaml
Comment (valerymo) [reply bundle/metadata/annotations.yaml:6]: ok, I see, it will be only in Community PR. Thanks

**Comment by eguzki** [comment]
> Are these changes a part of regular Community Release process?

Yes, those are changes to be in any release branch `3scale-2.X-stable` in upstream.

> What is the goal of current PR ? Is it to allow All tests to pass in temporary Testing PR https://github.com/3scale/3scale-operator/pull/950 ?

The goal is to have a (release) branch with the customization for the specific release. Any patch to 2.14.X should be done to this branch. And, as you say, it is good to run tests with all the customizations for the specific release.

>     Is it ok if Product Release branch origin/3scale-2.14-stable will have changes from Community release ?  Could merge of current PR impact on future Patch releases of 3scale-2.14 ?

`origin/3scale-2.14-stable ` is the base branch for any future release in community or downstream. Any future patch should go on top of these changes. For example, let's say there is a patch for backend and `quay.io/3scale/3scale214:apisonator-3scale-2.14.2-GA` is released, then we should update this branch with the new image and make sure tests pass. Then we decide if we want to cut a community release with the new patch. Same for downstream. The answer usually is patch releases only in product and for community, only minor releases. If we manage to automate the community release and do it small effort task, we could afford doing patch community releases. 

> Should we restore origin/3scale-2.14-stable branch - remove changes of Community? (after PR 950 pass)

No

> Just for notes - All related changes are already long time in Community release PR https://github.com/redhat-openshift-ecosystem/community-operators-prod/pull/4150 . So suppose that answer of quesion https://github.com/3scale/3scale-operator/issues/2 - Yes (?)

We want to have a good baseline in our `origin/3scale-2.14-stable`. So bundle info in `origin/3scale-2.14-stable` should be, theoretically, the same as the release in https://github.com/redhat-openshift-ecosystem/community-operators-prod/pull/4150. Today that is not possible, so we get as close as we can to the content going to https://github.com/redhat-openshift-ecosystem/community-operators-prod/pull/4150 and then we prepare a patch with the missing changes. Once #979  is merged, we bring changes to https://github.com/redhat-openshift-ecosystem/community-operators-prod/pull/4150 with the extra patch applied. That is why I told you from the beginning to focus on this repo's community bundle.



Hopefully that helps to understand

**Comment by eguzki** [comment]
All comments resolved, I think. But an approval is still pending. @MStokluska @valerymo

**Comment by valerymo** [comment]
> All comments resolved, I think. But an approval is still pending. @MStokluska @valerymo

Thanks very much for detailed reply on my questions regarding the Process.

**Comment by eguzki** [comment]
> /lgtm

I am happy to install any bot that will translate `/lgtm` to an approval. But currently we do not have that in this repo, so `/lgtm` is meaningless here.

**Comment by eguzki** [comment]
> /approve

I am happy to install any bot that will translate /approve to an approval. But currently we do not have that in this repo, so /approve is meaningless here.

**Comment by eguzki** [comment]
> /approval

I am happy to install any bot that will translate `/approval` to an approval. But currently we do not have that in this repo, so `/approval` is meaningless here.

**Review by MStokluska — APPROVED** [review]

**Comment by valerymo** [comment]
approval

**Pull request merged**
